### DISCUSSION TITLE
Copilot quality degrades on larger files

### DISCUSSION THREAD

**Discussion by GSPP** [community/community#9282] [Copilot Conversations] [answered]
I notice in a C# project that the larger a file gets the worse suggestions become. It's not just that context is missing. The suggestion can get a bit "stupid". It starts not understanding things that don't require much context. It gets to the point where I sometimes copy code out temporarily to a smaller file, get the suggestions there and copy back. I only do that when I need an important suggestion (e.g. a full binary search loop, pretty cool that it can do that 😄 )

It turns out that I routinely work on 1000 line files, and I believe that is not uncommon in other projects.

Just wanted to provide some feedback on that.

**Comment by GSPP** [comment]
Also, it sometimes seemed like quality was degrading based on time. Restarting VSCode seems to fix this. What could be the reason for this?

**Comment by flimtix** [comment]
I have exactly the same problem. As soon as I have a C# or Java project that has over 10 files (300 - 400 lines each) the quality becomes unacceptable.

**Comment by ShaunSHamilton** [comment]
+1 here. I find myself writing a lot of tests, and Copilot is excellent at helping with writing repetitive tests. It learns the patterns I have from earlier. However, my files end up being anywhere between 1000 and 6000 lines (_arguably, I could move a lot of the fixture data into a separate file, but I would still lose a lot of context_), and Copilot seems to 'forget' earlier used patterns.

I understand model inputs can degrade if over saturated, but I think it would be nice to have a config added to set (increase) an upper limit for number of lines read.

**Comment by davecheney** [comment, marked as answer]
@ShaunSHamilton OpenAI has a limit of around 1500 "tokens" for the prompt. A token is approximately 2-3 characters.

**Reply by GSPP** [reply]
@davecheney Maybe you guys can lobby for the limit to be increased?

Are there plans to improve context selection? I can imagine a number of ways this can be drastically improved. For example, with long context, function bodies can heuristically be emptied so that at least functions and signatures remain.

It is also conceivable to inject artificial context such as class names and member names from the project that seem relevant.

This seems like a field ripe for experimentation with all kinds of techniques.

**Reply by flimtix** [reply]
What if Copilot only looked at the current method and its references instead of loading entire files? Another solution would be to have it load only the current file for larger projects.

**Reply by cmp-nct** [reply]
The problem is the selection process, the plugin needs to do a better job in what code is presented to copilot for requests.
It should also include a "custom" section per file to ensure certain parts are present when requesting, like a SQL CREATE syntax for my main tables
I have files with 10000+ lines of code I am working on.

**Reply by CiberNin** [reply]
> What if Copilot only looked at the current method and its references instead of loading entire files? Another solution would be to have it load only the current file for larger projects.

Per recent decompilations of the copilot extension, it pulls only the most "relevant" 60 lines from each open file of the same language as the file you are currently editing. From a max of 20 different files. But in reality you will never get that much unless starting a new empty file because of the token budget.
